File: automata_test.py
__author__ = 'Alek'

import unittest

from dfa import Automata


class MyTestCase(unittest.TestCase):
    def test_automata_is_word_from_alphabet_method(self):
        symbols_list = ['a', 'b', 'c']
        a = Automata(symbols_list)
        self.assertEqual(a.is_word_from_alphabet('aaccabba'), True)
        self.assertEqual(a.is_word_from_alphabet('aacbdd'), False)
        single_symbol_word = 'a'
        self.assertEqual(a.is_word_from_alphabet(single_symbol_word), True)
        single_symbol_word = 'x'
        self.assertEqual(a.is_word_from_alphabet(single_symbol_word), False)

    def test_automata_compute_word_method_validation(self):
        symbols_list = ['a', 'b', 'c', '1', '5', '0']
        a = Automata(symbols_list)

        correct_word = 'aacc1ba0a'
        incorrect_word = 'aasdc'

        self.assertEqual(a.compute_word(incorrect_word), False)
        self.assertNotEqual(a.compute_word(correct_word), -1)
        incorrect_word = 'abc3'
        self.assertEqual(a.compute_word(incorrect_word), False)
        correct_word = '01005001'
        self.assertNotEquals(a.compute_word(correct_word), -1)

    """
    Automata.compute_word method tests
    """

    def test_binary_words_with_non_important_zeroes(self):
        """Binary word with non-important zeroes
            Should return  {Not_accepted} for '01001','00001'
            and {accepted} for '0', '101011'"""
        binary_symbols_list = '01'
        a = Automata(binary_symbols_list, 4)

        transition_lists = {
            '0': [1, 2, 2, 3],
            '1': [3, 2, 2, 3]
        }
        a.transition_lists = transition_lists
        accepting_states = [1, 3]

        incorrect_binary_words = ['01001', '00001', '010100010']
        correct_binary_words = ['0', '101011', '1010100010']

        for word in incorrect_binary_words:
            end_state = a.compute_word(word)
            self.assertNotIn(end_state, accepting_states)

        for word in correct_binary_words:
            end_state = a.compute_word(word)
            self.assertIn(end_state, accepting_states)

    def test_simple_regex_words(self):
        """First, a "b{0,1}a*b" regex language"""
        symbols = 'ab'
        a = Automata(symbols, 3)

        transition_lists = {
            'a': [1, 1, -1],
            'b': [1, 2, -1]
        }

        a.transition_lists = transition_lists
        accepting_states = [2]

        correct_words = ['baaaaaaab', 'ab', 'baab']
        incorrect_words = ['baaa', 'aaaa', 'b', 'a']

        for word in incorrect_words:
            end_state = a.compute_word(word)
            self.assertNotIn(end_state, accepting_states)

        for word in correct_words:
            end_state = a.compute_word(word)
            self.assertIn(end_state, accepting_states)


if __name__ == '__main__':
    unittest.main()

File: tests/automata_test.rb
require 'test/unit'

require_relative '../automata.rb'

class AutomataTest < Test::Unit::TestCase

  # Called before every test method runs. Can be used
  # to set up fixture information.
  def setup
    # Do nothing
  end

  # Called after every test method runs. Can be used to tear
  # down fixture information.

  def teardown
    # Do nothing
  end

  def test_is_word_from_alphabet
    symbols_list = %w(a b c)
    target = Automata.new(symbols_list, 2, nil)

    assert_equal(true, target.word_is_from_alphabet?('aaccabba'))
    assert_equal(false, target.word_is_from_alphabet?('aacbdd'))
  end

  def test_is_word_from_alphabet_single_symbol_word
    symbols_list = %w(a b c)
    target = Automata.new(symbols_list, 2, nil)

    assert_equal(true, target.word_is_from_alphabet?('a'))
    assert_equal(false, target.word_is_from_alphabet?('x'))
  end

  def test_automata_compute_word_method_validation
    symbols_list = %w(a b c 1 5 0)
    a = Automata.new(symbols_list, 2, nil)

    correct_word = 'aacc1ba0a'
    incorrect_word = 'aasdc'

    assert_equal(nil, a.compute_word(incorrect_word))
    assert_not_equal(-1, a.compute_word(correct_word))

    incorrect_word = 'abc3'
    assert_equal(nil, a.compute_word(incorrect_word))
    correct_word = '01005001'
    assert_not_equal(-1, a.compute_word(correct_word))
  end

  def test_binary_words_with_non_important_zeroes
  # Binary word with non-important zeroes
  #       Should return  {Not_accepted} for '01001','00001'
  #       and {accepted} for '0', '101011'
  binary_symbols_list = %w(0 1)
  transition_matrix = [[1, 2, 2, 3],
                       [3, 2, 2, 3]]
  a = Automata.new(binary_symbols_list, 4, transition_matrix)

  accepting_states = [1, 3]

  incorrect_binary_words = %w(01001 00001 010100010)
  correct_binary_words = %w(0 101011 1010100010)

  incorrect_binary_words.each do |word|
    end_state = a.compute_word(word)
    assert_equal(false, accepting_states.include?(end_state))
    end

    correct_binary_words.each do |word|
      end_state = a.compute_word(word)
      assert_equal(true, accepting_states.include?(end_state))
    end
  end

  def test_simple_regex_words
    # First, a "b{0,1}a*b" regex language
    symbols = %w(a b)
    transition_matrix = [[1, 1, -1],[1, 2, -1]]

    a = Automata.new(symbols, 3, transition_matrix)
    accepting_states = [2]
    correct_words = %w(baaaaaaab ab baab)
    incorrect_words = %w(baaa aaaa b a)

    incorrect_words.each do |word|
      end_state = a.compute_word(word)
      assert_equal(false, accepting_states.include?(end_state))
    end

    correct_words.each do |word|
      end_state = a.compute_word(word)
      assert_equal(true, accepting_states.include?(end_state))
    end
  end
end